File: example/matrix_multiply.cpp
#include <iostream>
#include <chrono>
#include <thread>
#include <vector>
#include <random>
#include <Eigen/Dense>
#include "xlnpwmon/xlnpwmon.h"

using namespace Eigen;
using namespace std;

/**
 * @brief Reduces matrix size to avoid memory issues
 */
constexpr int MATRIX_SIZE = 5000;
/**
 * @brief Number of threads to use for parallel processing
 */
constexpr int NUM_THREADS = 4;
/**
 * @brief Number of iterations for matrix multiplication
 */
constexpr int NUM_ITERATIONS = 10;

/**
 * @brief Structure to hold thread arguments
 */
struct ThreadArgs {
    int thread_id; /**< Thread identifier */
    int matrix_size; /**< Size of the matrix */
    int num_iterations; /**< Number of iterations */
};

/**
 * @brief Gets the current time in milliseconds
 * @return Current time in milliseconds
 */
double get_time_ms() {
    using namespace std::chrono;
    auto now = high_resolution_clock::now();
    return duration_cast<milliseconds>(now.time_since_epoch()).count();
}

/**
 * @brief Thread function for matrix multiplication
 * @param arg Pointer to thread arguments
 */
void* matrix_multiply_thread(void* arg) {
    auto* args = static_cast<ThreadArgs*>(arg);
    cout << "Thread " << args->thread_id << " starting...\n";

    // Eigen implementation
    MatrixXd A = MatrixXd::Random(args->matrix_size, args->matrix_size);
    MatrixXd B = MatrixXd::Random(args->matrix_size, args->matrix_size);
    MatrixXd C(args->matrix_size, args->matrix_size);

    for (int i = 0; i < args->num_iterations; i++) {
        C.noalias() = A * B;  // Using noalias() to avoid temporary variables
        A = C;  // Use result as input for next iteration
    }

    cout << "Thread " << args->thread_id << " completed\n";
    return nullptr;
}

int main() {
    pm_handle_t handle;
    pm_error_t error;
    pm_power_stats_t stats;
    vector<pthread_t> threads(NUM_THREADS);
    vector<ThreadArgs> thread_args(NUM_THREADS);
    double start_time, end_time;
    double total_time;

    /* Initialize the power monitor */
    error = pm_init(&handle);
    if (error != PM_SUCCESS) {
        cerr << "Failed to initialize power monitor: " << pm_error_string(error) << endl;
        return 1;
    }

    /* Start power sampling */
    error = pm_start_sampling(handle);
    if (error != PM_SUCCESS) {
        cerr << "Failed to start power sampling: " << pm_error_string(error) << endl;
        pm_cleanup(handle);
        return 1;
    }

    cout << "Starting power sampling...\n";

    /* Start matrix multiplication */
    cout << "Starting CPU-intensive task...\n";
    start_time = get_time_ms();

    /* Create and start threads */
    for (int i = 0; i < NUM_THREADS; i++) {
        thread_args[i].thread_id = i;
        thread_args[i].matrix_size = MATRIX_SIZE;
        thread_args[i].num_iterations = NUM_ITERATIONS;
        
        if (pthread_create(&threads[i], nullptr, matrix_multiply_thread, &thread_args[i]) != 0) {
            cerr << "Failed to create thread " << i << endl;
            pm_cleanup(handle);
            return 1;
        }
    }

    /* Wait for all threads to complete */
    for (int i = 0; i < NUM_THREADS; i++) {
        if (pthread_join(threads[i], nullptr) != 0) {
            cerr << "Failed to join thread " << i << endl;
        }
    }

    end_time = get_time_ms();
    total_time = (end_time - start_time) / 1000.0;
    cout << "CPU-intensive task completed\n";
    cout << "Total execution time: " << total_time << " seconds\n";

    /* Stop power sampling */
    error = pm_stop_sampling(handle);
    if (error != PM_SUCCESS) {
        cerr << "Failed to stop power sampling: " << pm_error_string(error) << endl;
        pm_cleanup(handle);
        return 1;
    }

    /* Get power statistics */
    error = pm_get_statistics(handle, &stats);
    if (error != PM_SUCCESS) {
        cerr << "Failed to get power statistics: " << pm_error_string(error) << endl;
        pm_cleanup(handle);
        return 1;
    }

    /* Print power statistics */
    cout << "\nPower Consumption Statistics:\n";
    cout << "Total Power Consumption:\n";
    cout << "  Minimum Value: " << stats.total.power.min << " W\n";
    cout << "  Maximum Value: " << stats.total.power.max << " W\n";
    cout << "  Average Value: " << stats.total.power.avg << " W\n";
    cout << "  Total Energy: " << stats.total.power.avg * total_time << " J\n";
    cout << "  Sample Count: " << stats.total.power.count << "\n";

    cout << "\nPer-Sensor Power Consumption Information:\n";
    for (int i = 0; i < stats.sensor_count; i++) {
        cout << "\nSensor: " << stats.sensors[i].name << "\n";
        cout << "  Minimum Value: " << stats.sensors[i].power.min << " W\n";
        cout << "  Maximum Value: " << stats.sensors[i].power.max << " W\n";
        cout << "  Average Value: " << stats.sensors[i].power.avg << " W\n";
        cout << "  Total Energy: " << stats.sensors[i].power.avg * total_time << " J\n";
        cout << "  Sample Count: " << stats.sensors[i].power.count << "\n";
    }

    /* Clean up */
    pm_cleanup(handle);
    return 0;
} 

File: bindings/rust/examples/matrix_multiply.rs
use xlnpwmon::PowerMonitor;
use std::thread;
use std::time::Duration;
use ndarray::{Array2, s};
use ndarray::linalg::Dot;
use rand::Rng;

/// Matrix size for multiplication
const MATRIX_SIZE: usize = 1000;
/// Number of threads to use for matrix multiplication
const NUM_THREADS: usize = 4;
/// Number of iterations for matrix multiplication
const NUM_ITERATIONS: usize = 10;

/// Executes matrix multiplication in a thread
fn matrix_multiply_thread(thread_id: usize) {
    println!("Thread {} starting execution...", thread_id);

    // Using ndarray for matrix multiplication
    let mut rng = rand::thread_rng();
    let mut a = Array2::zeros((MATRIX_SIZE, MATRIX_SIZE));
    let mut b = Array2::zeros((MATRIX_SIZE, MATRIX_SIZE));
    
    // Filling with random values
    for i in 0..MATRIX_SIZE {
        for j in 0..MATRIX_SIZE {
            a[[i, j]] = rng.gen_range(-1.0..1.0);
            b[[i, j]] = rng.gen_range(-1.0..1.0);
        }
    }

    // Performing multiple matrix multiplications
    let mut c = Array2::zeros((MATRIX_SIZE, MATRIX_SIZE));
    for _ in 0..NUM_ITERATIONS {
        c = a.dot(&b);
        a = c.clone();
    }

    println!("Thread {} completed execution", thread_id);
}

fn main() {
    println!("Xilinx Power Monitor - Matrix Multiplication Test");
    println!("=================================");

    // Initializing power monitor
    let monitor = PowerMonitor::new().unwrap();
    
    // Setting sampling frequency to 1000Hz
    monitor.set_sampling_frequency(1000).unwrap();
    
    // Resetting statistics
    monitor.reset_statistics().unwrap();
    
    // Starting sampling
    println!("Starting power sampling...");
    monitor.start_sampling().unwrap();
    
    // Recording start time
    let start_time = std::time::Instant::now();
    
    // Creating and starting threads
    let mut handles = vec![];
    for i in 0..NUM_THREADS {
        let handle = thread::spawn(move || {
            matrix_multiply_thread(i);
        });
        handles.push(handle);
    }
    
    // Waiting for all threads to complete
    for handle in handles {
        handle.join().unwrap();
    }
    
    // Calculating total execution time
    let total_time = start_time.elapsed().as_secs_f64();
    println!("\nTotal execution time: {:.2} seconds", total_time);
    
    // Waiting a short period to ensure data collection is complete
    thread::sleep(Duration::from_micros(500000));
    
    // Stopping sampling
    monitor.stop_sampling().unwrap();
    
    // Getting statistics
    let stats = monitor.get_statistics().unwrap();
    
    // Printing total power consumption statistics
    println!("\nPower Consumption Statistics:");
    println!("Total Power Consumption:");
    println!("  Minimum Value: {:.2} W", stats.total.power.min);
    println!("  Maximum Value: {:.2} W", stats.total.power.max);
    println!("  Average Value: {:.2} W", stats.total.power.avg);
    println!("  Total Energy Consumption: {:.2} J", stats.total.power.total);
    println!("  Sample Count: {}", stats.total.power.count);
    
    // Printing power consumption information for each sensor
    println!("\nPower Consumption Information for Each Sensor:");
    for i in 0..stats.sensor_count {
        let sensor = unsafe { &*stats.sensors.add(i as usize) };
        let name = String::from_utf8_lossy(&sensor.name).trim_matches('\0').to_string();
        println!("\nSensor: {}", name);
        println!("  Minimum Value: {:.2} W", sensor.power.min);
        println!("  Maximum Value: {:.2} W", sensor.power.max);
        println!("  Average Value: {:.2} W", sensor.power.avg);
        println!("  Total Energy Consumption: {:.2} J", sensor.power.total);
        println!("  Sample Count: {}", sensor.power.count);
    }
} 